src/main/java/com/openlattice/datastore/data/controllers/DataController.java

```diff
@@ -414,6 +417,51 @@ public Void clearEntitySet( UUID entitySetId ) {
         return null;
     }
 
+    @Override
+    @RequestMapping(
+            path = { "/" + ENTITY_SET + "/" + SET_ID_PATH + "/" + ENTITY_KEY_ID_PATH + "/graph" },
+            method = RequestMethod.DELETE
+    )
+    public Void clearEntityAndNeighborEntities(
+            @PathVariable( ENTITY_SET_ID ) UUID entitySetId,
+            @PathVariable( ENTITY_ID ) UUID entityId
+    ) {
+
+        ensureReadAccess( new AclKey( entitySetId ) );
+        Map<UUID, List<NeighborEntityDetails>> result = searchService
+                .executeEntityNeighborSearch( entitySetId, ImmutableSet.of( entityId ) );
+
+        if ( result != null && !result.containsKey( entityId ) ) {
+
+            List<NeighborEntityDetails> neighbors = result.get( entityId );
+            neighbors.parallelStream().forEach( neighbor -> {
+
+                UUID associationEntityId = (UUID) neighbor.getAssociationDetails().get( ID_FQN ).iterator().next();
+                UUID associationEntitySetId = neighbor.getAssociationEntitySet().getId();
+                Map<UUID, PropertyType> authorizedPropertyTypesOnNeighborAES = authzHelper
+                        .getAuthorizedPropertyTypes( associationEntitySetId, WRITE_PERMISSION );
+                dgm.clearEntities(
+                        neighbor.getAssociationEntitySet().getId(),
+                        ImmutableSet.of( associationEntityId ),
+                        authorizedPropertyTypesOnNeighborAES
+                );
+
+                if ( neighbor.getNeighborId().isPresent() && neighbor.getNeighborEntitySet().isPresent() ) {
+                    UUID neighborEntityId = neighbor.getNeighborId().get();
+                    UUID neighborEntitySetId = neighbor.getNeighborEntitySet().get().getId();
+                    Map<UUID, PropertyType> authorizedPropertyTypesOnNeighborES = authzHelper
```

Comment:
A lot of these auth checks are probably going to be repeated a bunch of times. It makes the code less pretty, but it would probably speed this up a lot to do an initial pass to check permissions, avoid repeating access checks, and then leverage the bulk delete. (I know for the PSA there are a bunch of people with hundreds of associated charges, so this would save thousands of unnecessary calls each time)

Reply:
I thought about that too. My plan was to get the basic logic down in the first implementation (this), and refactor after a few reviews.